axelor-production/src/main/java/com/axelor/apps/production/service/BillOfMaterialServiceImpl.java

```diff
@@ -83,24 +85,35 @@ public void updateProductCostPrice(BillOfMaterial billOfMaterial) throws AxelorE
 		billOfMaterialRepo.save(billOfMaterial);
 	}
 
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
+  public BillOfMaterial customizeBillOfMaterial(SaleOrderLine saleOrderLine)  {
 
-	@Override
-	@Transactional(rollbackOn = {AxelorException.class, Exception.class})
-	public BillOfMaterial customizeBillOfMaterial(SaleOrderLine saleOrderLine)  {
+    BillOfMaterial billOfMaterial = saleOrderLine.getBillOfMaterial();
+    return customizeBillOfMaterial(billOfMaterial);
 
-		BillOfMaterial billOfMaterial = saleOrderLine.getBillOfMaterial();
+  }
 
-		if(billOfMaterial != null)  {
-			BillOfMaterial personalizedBOM = JPA.copy(billOfMaterial, true);
-			billOfMaterialRepo.save(personalizedBOM);
-			personalizedBOM.setName(personalizedBOM.getName() + " ("+I18n.get(IExceptionMessage.BOM_1)+" " + personalizedBOM.getId() + ")");
-			personalizedBOM.setPersonalized(true);
-			return personalizedBOM;
-		}
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
+  public BillOfMaterial customizeBillOfMaterial(BillOfMaterial billOfMaterial) {
 
-		return null;
+    if(billOfMaterial != null)  {
+      BillOfMaterial personalizedBOM = JPA.copy(billOfMaterial, true);
+      billOfMaterialRepo.save(personalizedBOM);
+      personalizedBOM.setName(personalizedBOM.getName() + " ("+I18n.get(IExceptionMessage.BOM_1)+" " + personalizedBOM.getId() + ")");
+      personalizedBOM.setPersonalized(true);
+      Set<BillOfMaterial> personalizedBOMSet = new HashSet<BillOfMaterial>();
+      for(BillOfMaterial childBillOfMaterial : billOfMaterial.getBillOfMaterialSet()) {
+        personalizedBOMSet.add(customizeBillOfMaterial(childBillOfMaterial));
+      }
+      personalizedBOM.setBillOfMaterialSet(personalizedBOMSet);
 
-	}
+      return personalizedBOM;
+    }
+
+    return null;
+  }
```

Comment:
Please add a break condition to avoid to have an infinite loop if there is a cycle (max = 1000)